File: app/src/androidTest/java/com/orgzly/android/espresso/ShareActivityTest.java
package com.orgzly.android.espresso;

import android.content.Intent;
import androidx.test.rule.ActivityTestRule;

import com.orgzly.R;
import com.orgzly.android.AppIntent;
import com.orgzly.android.db.entity.NotePosition;
import com.orgzly.android.OrgzlyTest;
import com.orgzly.android.ui.share.ShareActivity;

import org.junit.Rule;
import org.junit.Test;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.orgzly.android.espresso.EspressoUtils.onSnackbar;
import static com.orgzly.android.espresso.EspressoUtils.toLandscape;
import static com.orgzly.android.espresso.EspressoUtils.toPortrait;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.startsWith;
import static org.junit.Assert.assertTrue;

//@Ignore
@SuppressWarnings("unchecked")
public class ShareActivityTest extends OrgzlyTest {
    @Rule
    public ActivityTestRule activityRule = new EspressoActivityTestRule<>(ShareActivity.class, true, false);

    private void startActivityWithIntent(String action, String type, String extraText, String searchQuery) {
        Intent intent = new Intent();

        if (action != null) {
            intent.setAction(action);
        }

        if (type != null) {
            intent.setType(type);
        }

        if (extraText != null) {
            intent.putExtra(Intent.EXTRA_TEXT, extraText);
        }

        if (searchQuery != null) {
            intent.putExtra(AppIntent.EXTRA_QUERY_STRING, searchQuery);
        }

        activityRule.launchActivity(intent);
    }

    @Test
    public void testDefaultBookRemainsSetAfterRotation() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);
        toPortrait(activityRule);
        onView(allOf(withId(R.id.fragment_note_location_button), isDisplayed()))
                .check(matches(withText(context.getString(R.string.default_share_notebook))));
        toLandscape(activityRule);
        onView(withId(R.id.fragment_note_location_button)).perform(scrollTo());
        onView(allOf(withId(R.id.fragment_note_location_button), isDisplayed()))
                .check(matches(withText(context.getString(R.string.default_share_notebook))));
    }

    @Test
    public void testBookRemainsSetAfterRotation() {
        testUtils.setupBook("book-one", "");
        testUtils.setupBook("book-two", "");
        testUtils.setupBook("book-three", "");
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);
        toPortrait(activityRule);
        onView(withId(R.id.fragment_note_location_button)).perform(click());
        onView(withText("book-two")).perform(click());
        onView(withId(R.id.fragment_note_location_button)).check(matches(withText("book-two")));
        toLandscape(activityRule);
        onView(withId(R.id.fragment_note_location_button)).check(matches(withText("book-two")));
    }

    @Test
    public void testDefaultBookName() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);

        onView(allOf(withId(R.id.fragment_note_location_button), isDisplayed()))
                .check(matches(withText(context.getString(R.string.default_share_notebook))));
    }

    @Test
    public void testTextSimple() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);
        onView(withId(R.id.done)).perform(click());
    }

    @Test
    public void testSaveAfterRotation() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);
        toLandscape(activityRule);
        toPortrait(activityRule);
        onView(withId(R.id.done)).perform(click());
    }

    @Test
    public void testTextEmpty() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "", null);
        onView(withId(R.id.done)).perform(click());
    }

    @Test
    public void testTextNull() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", null, null);
        onView(withId(R.id.done)).perform(click());
    }

    @Test
    public void testNoMatchingType() {
        startActivityWithIntent(Intent.ACTION_SEND, "application/octet-stream", null, null);

        onView(withId(R.id.fragment_note_title)).check(matches(withText("")));
        onSnackbar().check(matches(withText(context.getString(R.string.share_type_not_supported, "application/octet-stream"))));
    }

    @Test
    public void testNoActionSend() {
        startActivityWithIntent(null, null, null, null);

        onView(withId(R.id.fragment_note_title)).check(matches(withText("")));
    }

    @Test
    public void testSettingScheduledTimeRemainsSetAfterRotation() {
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", null);
        toPortrait(activityRule);
        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText("")));
        onView(withId(R.id.fragment_note_scheduled_button)).perform(click());
        onView(withText(R.string.set)).perform(click());
        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText(startsWith(defaultDialogUserDate()))));
        toLandscape(activityRule);
        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText(startsWith(defaultDialogUserDate()))));
    }

    @Test
    public void testNoteInsertedLast() {
        testUtils.setupBook("book-one", "* Note 1\n** Note 2");
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "Note 3", null);

        onView(withId(R.id.done)).perform(click());

        NotePosition n1 = dataRepository.getLastNote("Note 1").getPosition();
        NotePosition n2 = dataRepository.getLastNote("Note 2").getPosition();
        NotePosition n3 = dataRepository.getLastNote("Note 3").getPosition();

        assertTrue(n1.getLft() < n2.getLft());
        assertTrue(n2.getLft() < n2.getRgt());
        assertTrue(n2.getRgt() < n1.getRgt());
        assertTrue(n1.getRgt() < n3.getLft());
        assertTrue(n3.getLft() < n3.getRgt());
    }

    @Test
    public void testPresetBookFromSearchQuery() {
        testUtils.setupBook("foo", "doesn't matter");
        startActivityWithIntent(Intent.ACTION_SEND, "text/plain", "This is some shared text", "b.foo");

        onView(allOf(withId(R.id.fragment_note_location_button), isDisplayed()))
                .check(matches(withText("foo")));
    }
}

File: app/src/androidTest/java/com/orgzly/android/espresso/ShareActivityTest.kt
package com.orgzly.android.espresso

import android.content.Intent
import android.content.pm.ActivityInfo
import android.net.Uri
import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.scrollTo
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import com.orgzly.R
import com.orgzly.android.AppIntent
import com.orgzly.android.OrgzlyTest
import com.orgzly.android.espresso.EspressoUtils.*
import com.orgzly.android.prefs.AppPreferences
import com.orgzly.android.ui.share.ShareActivity
import org.hamcrest.Matchers.startsWith
import org.junit.Assert.assertTrue
import org.junit.Test


class ShareActivityTest : OrgzlyTest() {
    private fun startActivityWithIntent(
            action: String? = null,
            type: String? = null,
            extraText: String? = null,
            extraStreamUri: String? = null,
            queryString: String? = null): ActivityScenario<ShareActivity> {

        val intent = Intent(context, ShareActivity::class.java)

        if (action != null) {
            intent.action = action
        }

        if (type != null) {
            intent.type = type
        }

        if (extraText != null) {
            intent.putExtra(Intent.EXTRA_TEXT, extraText)
        }

        if (extraStreamUri != null) {
            intent.putExtra(Intent.EXTRA_STREAM, Uri.parse(extraStreamUri))
        }

        if (queryString != null) {
            intent.putExtra(AppIntent.EXTRA_QUERY_STRING, queryString)
        }

        return ActivityScenario.launch(intent)
    }

    @Test
    fun testDefaultBookRemainsSetAfterRotation() {
        val scenario = startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        onView(withId(R.id.fragment_note_location_button))
                .check(matches(withText(context.getString(R.string.default_share_notebook))))

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        }

        onView(withId(R.id.fragment_note_location_button))
                .check(matches(withText(context.getString(R.string.default_share_notebook))))
    }

    @Test
    fun testBookRemainsSetAfterRotation() {
        testUtils.setupBook("book-one", "")
        testUtils.setupBook("book-two", "")
        testUtils.setupBook("book-three", "")

        val scenario = startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        onView(withId(R.id.fragment_note_location_button)).perform(scrollTo(), click())
        onView(withText("book-two")).perform(click())
        onView(withId(R.id.fragment_note_location_button)).check(matches(withText("book-two")))

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        }

        onView(withId(R.id.fragment_note_location_button)).check(matches(withText("book-two")))
    }

    @Test
    fun testDefaultBookName() {
        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        onView(withId(R.id.fragment_note_location_button))
                .check(matches(withText(context.getString(R.string.default_share_notebook))))
    }

    @Test
    fun testTextSimple() {
        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testSaveAfterRotation() {
        val scenario = startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testTextEmpty() {
        startActivityWithIntent(action = Intent.ACTION_SEND, type = "text/plain", extraText = "")
        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testTextNull() {
        startActivityWithIntent(action = Intent.ACTION_SEND, type = "text/plain")
        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testImage() {
        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "image/png",
                extraStreamUri = "content://uri")

        onView(withId(R.id.fragment_note_title)).check(matches(withText("content://uri")))
        onView(withId(R.id.body_edit)).check(matches(withText("Cannot find image using this URI.")))

        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testFileCopy() {
        AppPreferences.attachMethod(context, ShareActivity.ATTACH_METHOD_COPY_DIR);
        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "application/pdf",
                extraStreamUri = "content://uri")

        onView(withId(R.id.fragment_note_title)).check(matches(withText("content://uri")))
        onView(withId(R.id.body_edit)).check(matches(withText("content://uri\n\nCannot determine fileName to this content.")))

        onView(withId(R.id.done)).perform(click())
    }

    @Test
    fun testNoMatchingType() {
        startActivityWithIntent(action = Intent.ACTION_SEND, type = "application/octet-stream")

        onView(withId(R.id.fragment_note_title)).check(matches(withText("")))
        onSnackbar().check(matches(withText(context.getString(R.string.share_type_not_supported, "application/octet-stream"))))
    }

    @Test
    fun testNoActionSend() {
        startActivityWithIntent()

        onView(withId(R.id.fragment_note_title)).check(matches(withText("")))
    }

    @Test
    fun testSettingScheduledTimeRemainsSetAfterRotation() {
        val scenario = startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text")

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        }

        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText("")))
        onView(withId(R.id.fragment_note_scheduled_button)).perform(click())
        onView(withText(R.string.set)).perform(click())
        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText(startsWith(defaultDialogUserDate()))))

        scenario.onActivity { activity ->
            activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        }

        onView(withId(R.id.fragment_note_scheduled_button)).check(matches(withText(startsWith(defaultDialogUserDate()))))
    }

    @Test
    fun testNoteInsertedLast() {
        testUtils.setupBook("book-one", "* Note 1\n** Note 2")

        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "Note 3")

        onView(withId(R.id.done)).perform(click())

        val (_, lft, rgt) = dataRepository.getLastNote("Note 1")!!.position
        val (_, lft1, rgt1) = dataRepository.getLastNote("Note 2")!!.position
        val (_, lft2, rgt2) = dataRepository.getLastNote("Note 3")!!.position

        assertTrue(lft < lft1)
        assertTrue(lft1 < rgt1)
        assertTrue(rgt1 < rgt)
        assertTrue(rgt < lft2)
        assertTrue(lft2 < rgt2)
    }

    @Test
    fun testPresetBookFromSearchQuery() {
        testUtils.setupBook("foo", "doesn't matter")

        startActivityWithIntent(
                action = Intent.ACTION_SEND,
                type = "text/plain",
                extraText = "This is some shared text",
                queryString = "b.foo")

        onView(withId(R.id.fragment_note_location_button)).check(matches(withText("foo")))
    }
}
